### ISSUE TITLE
Server cancellation isn't propagating to RPC

### ISSUE THREAD

**Issue by RReichert** [Tradias/asio-grpc#105]
# Description

I'm under the impression that a call to `grpc::Server::Shutdown()` would internally notify all currently registered and running awaitable async calls that the operation has been cancelled. That doesn't seem to be the case for me, what I've noticed is that the server blocks any new incoming requests and keeps the current existing requests uninformed of the shutdown.

Below is a sample code to demonstrate the issue, its a bi-direction steam server which prints any inbound message and periodically emits an outbound message. When I issue a `SIGINT` signal with a single client connected, the "read input" and "write output" messages continue to be printed for that existing client. If I close my client application, the server than finishes gracefully.

```cpp
int main(int argc, char *argv[]) {
  const auto port = argc >= 2 ? argv[1] : "5000";
  const auto host = std::string("0.0.0.0:") + port;

  std::unique_ptr<grpc::Server> server;

  grpc::ServerBuilder builder;
  agrpc::GrpcContext grpc_context{builder.AddCompletionQueue()};
  builder.AddListeningPort(host, grpc::InsecureServerCredentials());

  Service::AsyncService service;
  builder.RegisterService(&service);

  server = builder.BuildAndStart();

  ServerShutdown shutdown(*server, grpc_context);

  agrpc::register_awaitable_rpc_handler<RPC>(
      grpc_context, service,
      [&](RPC &rpc) -> boost::asio::awaitable<void> {
        const auto reader{[&]() -> boost::asio::awaitable<void> {
          Input input;
          while (co_await rpc.read(input)) {
            fprintf(stderr, "read input\n");
          }
        }};

        const auto writer{[&]() -> boost::asio::awaitable<void> {
          boost::asio::steady_timer timer(co_await boost::asio::this_coro::executor);
          Output output;

          while (co_await rpc.write(output)) {
            fprintf(stderr, "write output\n");

            timer.expires_after(std::chrono::seconds(1));
            co_await timer.async_wait(boost::asio::use_awaitable);
          }
        }};

        using namespace boost::asio::experimental::awaitable_operators;
        co_await (reader() || writer());
      },
      boost::asio::detached);

  grpc_context.run();
  fprintf(stderr, "Complete\n");
}
```

I'm not sure if I'm misunderstanding the gRPC framework or this is a case of the shutdown request not causing the `co_await rpc.write(output)` and `co_await rpc.read(input)` lines to return a `false` value. While trying to debug the issue, I've noticed that a call to `rpc.context().IsCancelled();` causes a segfault [here](https://github.com/grpc/grpc/blob/43e66cc29ab7b20a391cc17574b9692aae58dcab/src/core/lib/surface/completion_queue.cc#L1328), which I wasn't anticipating.

I'm currently running this on a Fedora 40 machine, compiled with Clang++ 18.1.6 with the following libraries:

* asio-grpc v3.1.0
* boost 1.83.0
* grpc 1.45.2 (used later ones like grpc 1.63.0, same issue)

**Comment by Tradias** [comment]
> what I've noticed is that the server blocks any new incoming requests and keeps the current existing requests uninformed of the shutdown.

Correct, the [official gRPC documentation](https://grpc.github.io/grpc/cpp/namespacegrpc.html) on [Server::Shutdown](https://grpc.github.io/grpc/cpp/classgrpc_1_1_server_interface.html#a644772d11101318f1424e595dee73ccf) provides more insights. You could try the Shutdown-with-timeout method and see whether it helps.

> While trying to debug the issue, I've noticed that a call to rpc.context().IsCancelled(); causes a segfault

Again the [official documentation](https://grpc.github.io/grpc/cpp/classgrpc_1_1_server_context.html#af2d0f087805b4b475d01b12d73508f09) explains why: 

> IsCancelled is always safe to call when using sync or callback API. When using async API, it is only safe to call IsCancelled after the AsyncNotifyWhenDone tag has been delivered. Thread-safe.

In asio-grpc terms you need to make sure that `rpc.wait_for_done()` has completed before calling `IsCancelled`. I have noticed that there is no example code for that, I will add one.

**Comment by RReichert** [comment]
Appreciate the help. It my mistake here not noticing that the `Shutdown` had an overloaded templated method which took in the time point. Using that did the trick.

Also appreciate the clarification around the `IsCancelled`. I'll close this issue off.

**Comment by Tradias** [comment]
Great, I am glad that worked!

In case you are interested there is now also an example on how `rpc.wait_for_done()` could be used: https://github.com/Tradias/asio-grpc/blob/master/example/streaming-server.cpp#L96-L144
